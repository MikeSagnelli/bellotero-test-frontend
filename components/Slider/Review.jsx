import React from 'react';
import PropTypes from 'prop-types';
import {
  ReviewContainer,
  AuthorContainer,
  CommentContainer,
  AuthorName,
  AuthorPosition,
  Comment
} from './styles';

const Review = ({ review }) => (
  <ReviewContainer>
    <AuthorContainer>
      <AuthorName id="author-name">{review.name}</AuthorName>
      <AuthorPosition id="author-position">{review.position}</AuthorPosition>
    </AuthorContainer>
    <CommentContainer>
      <Comment id="author-comment">{review.comment}</Comment>
    </CommentContainer>
  </ReviewContainer>
);

Review.displayName = 'Review';

Review.propTypes = {
  review: PropTypes.shape({
    name: PropTypes.string,
    position: PropTypes.string,
    comment: PropTypes.string
  }).isRequired
};

export default Review;
