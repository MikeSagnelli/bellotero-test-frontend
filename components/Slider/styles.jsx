import styled from 'styled-components';
import { primaryColor, whiteColor, textColor, lightTextColor } from '../../static/styleConstants';

export const TitleContainer = styled.div`
  margin-top: -20%;
  width: 389px;
  height: 56px;
  background-color: ${primaryColor};
`;

export const TitleText = styled.h1`
  margin: 8px 4px;
  font-family: 'Roboto', sans-serif;
  font-size: 36px;
  font-weight: 900;
  line-height: 1.11;
  letter-spacing: 0.8px;
  color: ${whiteColor};
`;

export const ReviewContainer = styled.div`
  display: flex;
  margin-top: 80px;
  width: calc(100% - 78px);
  background-color: ${whiteColor};
  min-height: 254px;
  height: fit-content;
`;

export const AuthorContainer = styled.div`
  width: 40%;
  padding: 32px;
`;

export const AuthorName = styled.h2`
  font-family: Roboto;
  font-size: 32px;
  font-weight: bold;
  line-height: 1.5;
  color: ${textColor};
  margin: 0;
`;

export const AuthorPosition = styled.h3`
  font-family: Roboto;
  font-size: 14px;
  font-weight: normal;
  line-height: normal;
  color: ${lightTextColor};
  margin: 0;
`;

export const CommentContainer = styled.div`
  width: 60%;
  padding: 32px;
  height: fit-content;
`;

export const Comment = styled.p`
  font-family: Roboto;
  font-size: 24px;
  font-weight: bold;
  line-height: 1.33;
  color: ${textColor};
  margin: 0;
`;

export const ControlsContainer = styled.div`
  display: flex;
  justify-content: flex-end;
  margin-top: -28px;
  width: 100%;
  height: fit-content;
`;

export const CounterContainer = styled.div`
  width: 120px;
  height: 56px;
  background-color: ${primaryColor};
  margin-right: 1px;
`;

export const CounterText = styled.h2`
  width: 35px;
  height: 32px;
  font-family: 'Cormorant Garamond', serif;
  font-size: 32px;
  font-weight: 600;
  font-style: italic;
  line-height: 1;
  letter-spacing: 0.7px;
  color: ${whiteColor};
  margin: 8px 44px 16px 41px;
`;

export const SwitchersContainer = styled.div`
  width: fit-content;
  height: fit-content;
  display: flex;
`;

export const SwitchButton = styled.div`
  width: 56px;
  height: 56px;
  background-color: ${primaryColor};
  cursor: pointer;
`;

export const Arrow = styled.img`
  width: 16px;
  height: 18px;
  margin: 20px 20px 19px 20px;
`;

export const RightArrow = styled(Arrow)`
  transform: rotate(180deg);
`;
