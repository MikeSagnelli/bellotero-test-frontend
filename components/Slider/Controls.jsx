import React from 'react';
import PropTypes from 'prop-types';
import { Arrow, RightArrow, SwitchersContainer, SwitchButton } from './styles';

const Counter = ({ previousReview, nextReview }) => (
  <SwitchersContainer>
    <SwitchButton id="prev-button" onClick={previousReview}>
      <Arrow src="/static/images/arrow.svg" alt="next" />
    </SwitchButton>
    <SwitchButton id="next-button" onClick={nextReview}>
      <RightArrow src="/static/images/arrow.svg" alt="next" />
    </SwitchButton>
  </SwitchersContainer>
);

Counter.displayName = 'Counter';

Counter.propTypes = {
  previousReview: PropTypes.func.isRequired,
  nextReview: PropTypes.func.isRequired
};

export default Counter;
