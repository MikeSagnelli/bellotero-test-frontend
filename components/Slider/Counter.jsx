import React from 'react';
import PropTypes from 'prop-types';
import { CounterContainer, CounterText } from './styles';

const Counter = ({ currentReview, totalReviews }) => (
  <CounterContainer>
    <CounterText id="current-review">{`${currentReview}/${totalReviews}`}</CounterText>
  </CounterContainer>
);

export default Counter;

Counter.displayName = 'Counter';

Counter.propTypes = {
  currentReview: PropTypes.number.isRequired,
  totalReviews: PropTypes.number.isRequired
};
