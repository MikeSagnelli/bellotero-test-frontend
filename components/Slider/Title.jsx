import React from 'react';
import PropTypes from 'prop-types';
import { TitleContainer, TitleText } from './styles';

const Title = ({ title }) => (
  <TitleContainer>
    <TitleText id="title">{title}</TitleText>
  </TitleContainer>
);

export default Title;

Title.displayName = 'Title';

Title.propTypes = {
  title: PropTypes.string.isRequired
};
