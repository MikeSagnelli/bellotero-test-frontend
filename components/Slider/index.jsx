import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Title from './Title';
import Review from './Review';
import Counter from './Counter';
import Controls from './Controls';
import { ControlsContainer } from './styles';

class Slider extends React.PureComponent {
  state = {
    currentReview: 0,
    totalReviews: 0
  };

  componentDidMount() {
    const { testimonialData } = this.props;

    this.setState({ totalReviews: testimonialData.slider.reviews.length });
  }

  nextReview = () => {
    const { currentReview, totalReviews } = this.state;

    if (currentReview < totalReviews - 1) {
      this.setState({ currentReview: currentReview + 1 });
    }
  };

  previousReview = () => {
    const { currentReview } = this.state;

    if (currentReview > 0) {
      this.setState({ currentReview: currentReview - 1 });
    }
  };

  render() {
    const { testimonialData } = this.props;
    const { currentReview } = this.state;

    return (
      <React.Fragment>
        <Title title={testimonialData.slider.title} />
        <Review review={testimonialData.slider.reviews[currentReview]} />
        <ControlsContainer>
          <Counter
            currentReview={currentReview + 1}
            totalReviews={testimonialData.slider.reviews.length}
          />
          <Controls previousReview={this.previousReview} nextReview={this.nextReview} />
        </ControlsContainer>
      </React.Fragment>
    );
  }
}

Slider.displayName = 'Slider';

Slider.propTypes = {
  testimonialData: PropTypes.shape({
    slider: PropTypes.shape({
      title: PropTypes.string,
      reviews: PropTypes.arrayOf(
        PropTypes.shape({
          name: PropTypes.string,
          position: PropTypes.string,
          comment: PropTypes.string
        })
      )
    })
  })
};

Slider.defaultProps = {
  testimonialData: null
};

export default connect(state => state)(Slider);
