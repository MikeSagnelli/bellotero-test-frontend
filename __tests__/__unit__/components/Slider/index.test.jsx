import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import wait from 'waait';
import Slider from '../../../../components/Slider';

const mockStore = configureStore();

const initialState = {
  testimonialData: {
    slider: {
      title: 'Our customers love us',
      reviews: [
        {
          name: 'Pete Zahut',
          position: 'Chief @ Maniak',
          comment:
            "It's funny what memory does, isn't it? My favorite holiday tradition might not have happened more than once or twice. But because it is such a good memory, so encapsulating of everything I love about the holidays, in my mind it happened every year. Without fail"
        },
        {
          name: 'Bernabe',
          position: 'Tech Lead @ Maniak',
          comment:
            'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut sed lectus quam. Curabitur ultricies pretium orci nec finibus. Nullam congue quis tortor a tempus. Morbi rutrum, nunc at hendrerit gravida, tortor turpis molestie nibh, vel varius augue ante eu velit.'
        }
      ]
    }
  }
};

describe('Slider', () => {
  it('renders slider', () => {
    const component = shallow(<Slider />);
    expect(component).toMatchSnapshot();
  });

  it('renders title correctly', async () => {
    const store = mockStore(initialState);

    const component = mount(
      <Provider store={store}>
        <Slider />
      </Provider>
    );

    await wait(0);
    component.update();

    const title = component.find('h1#title').text();
    expect(title).toBe(initialState.testimonialData.slider.title);

    component.unmount();
  });

  it('renders first info correctly', async () => {
    const store = mockStore(initialState);

    const component = mount(
      <Provider store={store}>
        <Slider />
      </Provider>
    );

    await wait(0);
    component.update();

    const authorName = component.find('h2#author-name').text();
    const authorPosition = component.find('h3#author-position').text();
    const authorComment = component.find('p#author-comment').text();
    expect(authorName).toBe(initialState.testimonialData.slider.reviews[0].name);
    expect(authorPosition).toBe(initialState.testimonialData.slider.reviews[0].position);
    expect(authorComment).toBe(initialState.testimonialData.slider.reviews[0].comment);

    component.unmount();
  });

  it('renders first info correctly, then when clicking renders next info', async () => {
    const store = mockStore(initialState);

    const component = mount(
      <Provider store={store}>
        <Slider />
      </Provider>
    );

    await wait(0);
    component.update();

    let authorName;
    let authorPosition;
    let authorComment;

    authorName = component.find('h2#author-name').text();
    authorPosition = component.find('h3#author-position').text();
    authorComment = component.find('p#author-comment').text();
    const nextButton = component.find('div#next-button');

    expect(authorName).toBe(initialState.testimonialData.slider.reviews[0].name);
    expect(authorPosition).toBe(initialState.testimonialData.slider.reviews[0].position);
    expect(authorComment).toBe(initialState.testimonialData.slider.reviews[0].comment);

    nextButton.simulate('click');

    await wait(0);
    component.update();

    authorName = component.find('h2#author-name').text();
    authorPosition = component.find('h3#author-position').text();
    authorComment = component.find('p#author-comment').text();

    expect(authorName).toBe(initialState.testimonialData.slider.reviews[1].name);
    expect(authorPosition).toBe(initialState.testimonialData.slider.reviews[1].position);
    expect(authorComment).toBe(initialState.testimonialData.slider.reviews[1].comment);

    component.unmount();
  });

  it('renders first info correctly, then when clicking renders next info, then goes back', async () => {
    const store = mockStore(initialState);

    const component = mount(
      <Provider store={store}>
        <Slider />
      </Provider>
    );

    await wait(0);
    component.update();

    let authorName;
    let authorPosition;
    let authorComment;

    authorName = component.find('h2#author-name').text();
    authorPosition = component.find('h3#author-position').text();
    authorComment = component.find('p#author-comment').text();
    const nextButton = component.find('div#next-button');
    const prevButton = component.find('div#prev-button');

    expect(authorName).toBe(initialState.testimonialData.slider.reviews[0].name);
    expect(authorPosition).toBe(initialState.testimonialData.slider.reviews[0].position);
    expect(authorComment).toBe(initialState.testimonialData.slider.reviews[0].comment);

    nextButton.simulate('click');

    await wait(0);
    component.update();

    authorName = component.find('h2#author-name').text();
    authorPosition = component.find('h3#author-position').text();
    authorComment = component.find('p#author-comment').text();

    expect(authorName).toBe(initialState.testimonialData.slider.reviews[1].name);
    expect(authorPosition).toBe(initialState.testimonialData.slider.reviews[1].position);
    expect(authorComment).toBe(initialState.testimonialData.slider.reviews[1].comment);

    prevButton.simulate('click');

    await wait(0);
    component.update();

    authorName = component.find('h2#author-name').text();
    authorPosition = component.find('h3#author-position').text();
    authorComment = component.find('p#author-comment').text();

    expect(authorName).toBe(initialState.testimonialData.slider.reviews[0].name);
    expect(authorPosition).toBe(initialState.testimonialData.slider.reviews[0].position);
    expect(authorComment).toBe(initialState.testimonialData.slider.reviews[0].comment);

    component.unmount();
  });

  it('renders first info correctly, then when clicking renders next info, then tries to click next and stays the same', async () => {
    const store = mockStore(initialState);

    const component = mount(
      <Provider store={store}>
        <Slider />
      </Provider>
    );

    await wait(0);
    component.update();

    let authorName;
    let authorPosition;
    let authorComment;

    authorName = component.find('h2#author-name').text();
    authorPosition = component.find('h3#author-position').text();
    authorComment = component.find('p#author-comment').text();
    const nextButton = component.find('div#next-button');

    expect(authorName).toBe(initialState.testimonialData.slider.reviews[0].name);
    expect(authorPosition).toBe(initialState.testimonialData.slider.reviews[0].position);
    expect(authorComment).toBe(initialState.testimonialData.slider.reviews[0].comment);

    nextButton.simulate('click');

    await wait(0);
    component.update();

    authorName = component.find('h2#author-name').text();
    authorPosition = component.find('h3#author-position').text();
    authorComment = component.find('p#author-comment').text();

    expect(authorName).toBe(initialState.testimonialData.slider.reviews[1].name);
    expect(authorPosition).toBe(initialState.testimonialData.slider.reviews[1].position);
    expect(authorComment).toBe(initialState.testimonialData.slider.reviews[1].comment);

    nextButton.simulate('click');

    await wait(0);
    component.update();

    authorName = component.find('h2#author-name').text();
    authorPosition = component.find('h3#author-position').text();
    authorComment = component.find('p#author-comment').text();

    expect(authorName).toBe(initialState.testimonialData.slider.reviews[1].name);
    expect(authorPosition).toBe(initialState.testimonialData.slider.reviews[1].position);
    expect(authorComment).toBe(initialState.testimonialData.slider.reviews[1].comment);

    component.unmount();
  });

  it('renders first info correctly, then when clicking previous it stays the same', async () => {
    const store = mockStore(initialState);

    const component = mount(
      <Provider store={store}>
        <Slider />
      </Provider>
    );

    await wait(0);
    component.update();

    let authorName;
    let authorPosition;
    let authorComment;

    authorName = component.find('h2#author-name').text();
    authorPosition = component.find('h3#author-position').text();
    authorComment = component.find('p#author-comment').text();
    const prevButton = component.find('div#prev-button');

    expect(authorName).toBe(initialState.testimonialData.slider.reviews[0].name);
    expect(authorPosition).toBe(initialState.testimonialData.slider.reviews[0].position);
    expect(authorComment).toBe(initialState.testimonialData.slider.reviews[0].comment);

    prevButton.simulate('click');

    await wait(0);
    component.update();

    authorName = component.find('h2#author-name').text();
    authorPosition = component.find('h3#author-position').text();
    authorComment = component.find('p#author-comment').text();

    expect(authorName).toBe(initialState.testimonialData.slider.reviews[0].name);
    expect(authorPosition).toBe(initialState.testimonialData.slider.reviews[0].position);
    expect(authorComment).toBe(initialState.testimonialData.slider.reviews[0].comment);

    component.unmount();
  });
});
