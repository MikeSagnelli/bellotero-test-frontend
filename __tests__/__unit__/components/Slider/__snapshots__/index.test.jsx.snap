// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders slider 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
